rust/src/shards/gui/containers/docking.rs

```diff
@@ -0,0 +1,460 @@
+    util::require_parents(&mut self.requiring, &self.parents);
+
+    Some(&self.requiring)
+  }
+
+  fn exposedVariables(&mut self) -> Option<&ExposedTypes> {
+    self.exposing.clear();
+
+    if util::expose_contents_variables(&mut self.exposing, &self.contents) {
+      Some(&self.exposing)
+    } else {
+      None
+    }
+  }
+
+  fn hasCompose() -> bool {
+    true
+  }
+
+  fn compose(&mut self, data: &InstanceData) -> Result<Type, &str> {
+    if !self.contents.is_empty() {
+      self.contents.compose(&data)?;
+    }
+
+    // Always passthrough the input
+    Ok(data.inputType)
+  }
+
+  fn warmup(&mut self, ctx: &Context) -> Result<(), &str> {
+    self.parents.warmup(ctx);
+    self.title.warmup(ctx);
+    if !self.contents.is_empty() {
+      self.contents.warmup(ctx)?;
+    }
+
+    Ok(())
+  }
+
+  fn cleanup(&mut self) -> Result<(), &str> {
+    if !self.contents.is_empty() {
+      self.contents.cleanup();
+    }
+    self.title.cleanup();
+    self.parents.cleanup();
+
+    Ok(())
+  }
+
+  fn activate(&mut self, context: &Context, input: &Var) -> Result<Var, &str> {
+    if self.contents.is_empty() {
+      // no contents, same as inactive
+      return Ok(false.into());
+    }
+
+    if let Some(ui) = util::get_current_parent(*self.parents.get())? {
+      util::activate_ui_contents(context, input, ui, &mut self.parents, &mut self.contents)?;
+
+      // Always passthrough the input
+      Ok(*input)
+    } else {
+      Err("No UI parent")
+    }
+  }
+}
+
+impl Default for DockArea {
+  fn default() -> Self {
+    let mut ctx = ParamVar::default();
+    ctx.set_name(CONTEXTS_NAME);
+    let mut parents = ParamVar::default();
+    parents.set_name(PARENTS_UI_NAME);
+    Self {
+      instance: ctx,
+      requiring: Vec::new(),
+      contents: ParamVar::default(),
+      parents,
+      exposing: Vec::new(),
+      headers: Vec::new(),
+      shards: Vec::new(),
+      tabs: egui_dock::Tree::new(Vec::new()),
+    }
+  }
+}
+
+impl Shard for DockArea {
+  fn registerName() -> &'static str
+  where
+    Self: Sized,
+  {
+    cstr!("UI.DockArea")
+  }
+
+  fn hash() -> u32
+  where
+    Self: Sized,
+  {
+    compile_time_crc32::crc32!("UI.DockArea-rust-0x20200101")
+  }
+
+  fn name(&mut self) -> &str {
+    "UI.DockArea"
+  }
+
+  fn help(&mut self) -> OptionalString {
+    OptionalString(shccstr!("TODO."))
+  }
+
+  fn inputTypes(&mut self) -> &Types {
+    &ANY_TYPES
+  }
+
+  fn inputHelp(&mut self) -> OptionalString {
+    OptionalString(shccstr!("TODO."))
+  }
+
+  fn outputTypes(&mut self) -> &Types {
+    &ANY_TYPES
+  }
+
+  fn outputHelp(&mut self) -> OptionalString {
+    OptionalString(shccstr!("TODO."))
+  }
+
+  fn parameters(&mut self) -> Option<&Parameters> {
+    Some(&DOCKAREA_PARAMETERS)
+  }
+
+  fn setParam(&mut self, index: i32, value: &Var) -> Result<(), &str> {
+    match index {
+      0 => {
+        let seq = Seq::try_from(value)?;
+        for tab_spec in seq.iter().rev().filter_map(|s| {
```

Comment:
Can you maybe assign the result of filter_map to a variable to make it more readable?

Reply:
I don't think it makes it more readable since it is an iterator. It's quite idiomatic to use a for loop on an iterator.